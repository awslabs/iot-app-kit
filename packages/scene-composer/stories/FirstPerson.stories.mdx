import { Canvas, Meta, Story } from '@storybook/addon-docs';

import SceneComposer, { argTypes, args } from './components/scene-composer';
import { testScenes } from '../tests/testData';
import { COMPOSER_FEATURES } from '../src/interfaces/feature';

<Meta title="Tests/FirstPerson" component={SceneComposer} argTypes={argTypes} />

export const defaultArgs = {
  source: 'local',
  scene: 'scene1',
  theme: 'dark',
  mode: 'Viewing',
  density: 'comfortable',
  features: [
    COMPOSER_FEATURES.FirstPerson,
  ]
}

export const Template = (args) => {
  return <SceneComposer {...defaultArgs} {...args} />
}

# Local Viewing

<Canvas>
  <Story name='Local Viewer'
    height='500px'
    parameters={{ controls: { include: ['onSelectionChanged', 'onWidgetClick'], hideNoControlsWarning: true }}}
    args={{ ...defaultArgs, scene: 'CookieFactoryWaterTank' }}>
    {Template.bind({})}
  </Story>
</Canvas>

# AWS Viewing

<Canvas>
  <Story name='AWS Viewer'
    height='500px'
    args={{ ...defaultArgs, source: 'aws', }}>
    {Template.bind({})}
  </Story>
</Canvas>
