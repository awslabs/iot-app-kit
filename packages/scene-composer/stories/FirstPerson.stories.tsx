import type { Meta, StoryObj } from '@storybook/react';
import { COMPOSER_FEATURES } from '../src/interfaces/feature';
import SceneComposer, { argTypes } from './components/scene-composer';

export default {
  title: 'Tests/FirstPerson',
  component: SceneComposer,
  argTypes,
  args: {
    source: 'local',
    scene: 'scene1',
    // @ts-expect-error type is wrong?
    assetType: 'GLB',
    theme: 'dark',
    mode: 'Viewing',
    density: 'comfortable',
    features: [COMPOSER_FEATURES.FirstPerson],
  },
} satisfies Meta<typeof SceneComposer>;

type Story = StoryObj<typeof SceneComposer>;

export const LocalViewer: Story = {
  parameters: {
    docs: {
      inlineStories: false,
      iframeHeight: '500px',
    },
    controls: {
      include: ['onSelectionChanged', 'onWidgetClick'],
      hideNoControlsWarning: true,
    },
  },
};

export const AWSViewer: Story = {
  parameters: {
    docs: {
      inlineStories: false,
      iframeHeight: '500px',
    },
  },
  args: {
    source: 'aws',
  },
};
