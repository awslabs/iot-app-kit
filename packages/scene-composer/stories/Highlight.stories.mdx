import { Canvas, Meta, Story } from '@storybook/addon-docs';

import { argTypes, args } from './components/scene-viewer';
import HighlightTestWrapper from  './components/highlightTestWrapper';
import { testScenes } from '../tests/testData';
import { COMPOSER_FEATURES } from '../src/interfaces/feature';

<Meta title="Tests/Scene Viewer" component={HighlightTestWrapper} argTypes={argTypes} />

export const defaultArgs = {
  source: 'local',
  scene: 'scene1',
  theme: 'dark',
  mode: 'Viewing',
  density: 'comfortable',
  features: []
}

export const Template = (args) => {
  return <HighlightTestWrapper {...defaultArgs} {...args} />
}

# Scene Viewer

## Highlights

### Verification

Motion Indicators should disappear when turned off.

#### Steps:
- Click the "View Options" Menu
- Turn off Motion Indicators

#### Verification:
- Motion Indicator should not be visible in the scene

<Canvas>
  <Story name='Highlights'
    height='500px'
    parameters={{ controls: { include: ['onSelectionChanged', 'onWidgetClick'], hideNoControlsWarning: true }}}
    args={{ ...defaultArgs, scene: 'CookieFactoryWaterTank' }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Sub Model Selection
