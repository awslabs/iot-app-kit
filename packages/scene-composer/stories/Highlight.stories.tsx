import type { Meta, StoryObj } from '@storybook/react';
import { COMPOSER_FEATURES } from '../src/interfaces/feature';
import HighlightTestWrapper from './components/highlightTestWrapper';
import { argTypes } from './components/scene-viewer';

export default {
  title: 'Tests/Scene Viewer Highlight',
  component: HighlightTestWrapper,
  argTypes,
  args: {
    source: 'local',
    scene: 'scene1',
    // @ts-expect-error type is wrong?
    assetType: 'GLB',
    theme: 'dark',
    mode: 'Viewing',
    density: 'comfortable',
    features: [COMPOSER_FEATURES.FirstPerson],
  },
} satisfies Meta<typeof HighlightTestWrapper>;

type Story = StoryObj<typeof HighlightTestWrapper>;

export const Highlights: Story = {
  parameters: {
    docs: { inlineStories: false, iframeHeight: '500px' },
    controls: { include: ['onSelectionChanged', 'onWidgetClick'], hideNoControlsWarning: true },
  },
  args: {
    scene: 'CookieFactoryWaterTank',
  },
};
