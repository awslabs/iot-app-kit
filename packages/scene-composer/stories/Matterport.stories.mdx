import { Canvas, Meta, Story } from '@storybook/addon-docs';

import SceneComposer, { argTypes, args } from './components/scene-composer';
import { testScenes } from '../tests/testData';
import { COMPOSER_FEATURES } from '../src/interfaces/feature';

<Meta title="Tests/Matterport" component={SceneComposer} argTypes={argTypes} />

export const defaultArgs = {
  source: 'local',
  scene: 'scene1',
  theme: 'dark',
  mode: 'Viewing',
  density: 'comfortable',
  matterportModelId: 'cZowU1FxWEQ',
  matterportApplicationKey: '',
  features: [
    COMPOSER_FEATURES.SceneHierarchyReorder,
    COMPOSER_FEATURES.SceneHierarchySearch,
    COMPOSER_FEATURES.SubModelSelection,
    COMPOSER_FEATURES.ENHANCED_EDITING,
    COMPOSER_FEATURES.CameraView,
    COMPOSER_FEATURES.OpacityRule,
    COMPOSER_FEATURES.Matterport
  ]
}

export const Template = (args) => {
  return <SceneComposer {...defaultArgs} {...args} />
}

# Viewing

<Canvas>
  <Story name='Viewer'
    height='500px'
    parameters={{ controls: { include: ['matterportModelId', 'matterportApplicationKey', 'onSelectionChanged', 'onWidgetClick'], hideNoControlsWarning: true }}}
    args={{ ...defaultArgs, scene: 'CookieFactoryWaterTank' }}>
    {Template.bind({})}
  </Story>
</Canvas>

# Editing
<Canvas>
  <Story name='Editor'
    height='500px'
    parameters={{ controls: { include: ['matterportModelId', 'matterportApplicationKey', 'onSelectionChanged', 'onWidgetClick'], hideNoControlsWarning: true }}}
    args={{ ...defaultArgs, scene: 'CookieFactoryWaterTankMatterportTag', mode: 'Editing' }}>
    {Template.bind({})}
  </Story>
</Canvas>
