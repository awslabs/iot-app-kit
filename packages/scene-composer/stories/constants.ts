import { FIELD_NAME } from './types';

export const entityIds = Object.seal(['room1', 'room2']);
export const entityNames = Object.seal(['room1Name', 'room2Name']);
export const allEntityIds = Object.seal([...entityIds, 'room3', 'room4']);
export const componentNames = Object.seal(['temperatureSensor1', 'temperatureSensor2']);
export const propertyNames = Object.seal(['temperature']);

export const FIELDS: FIELD_NAME[] = ['entityId', 'componentName', 'propertyName'];

export const MOCK_DELAY = 300;
