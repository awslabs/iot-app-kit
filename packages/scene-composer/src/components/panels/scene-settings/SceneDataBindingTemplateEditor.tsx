import React, { useCallback, useContext, useEffect, useMemo, useState } from 'react';
import { useIntl } from 'react-intl';
import { cloneDeep } from 'lodash';
import { FormField, Select } from '@cloudscape-design/components';

import useLifecycleLogging from '../../../logger/react-logger/hooks/useLifecycleLogging';
import { EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE } from '../../../common/constants';
import { accessStore } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import {
  IDataBindingConfig,
  IDataFieldOption,
  IValueDataBinding,
  IValueDataBindingProvider,
  IValueDataBindingProviderState,
  KnownSceneProperty,
} from '../../../interfaces';
import { dataBindingConfigSelector } from '../../../utils/dataBindingTemplateUtils';

interface SceneDataBindingTemplateEditorProps {
  valueDataBindingProvider: IValueDataBindingProvider;
}

export const SceneDataBindingTemplateEditor: React.FC<SceneDataBindingTemplateEditorProps> = ({
  valueDataBindingProvider,
}) => {
  useLifecycleLogging('SceneDataBindingTemplateEditor');

  const sceneComposerId = useContext(sceneComposerIdContext);
  const intl = useIntl();
  const { setSceneProperty } = accessStore(sceneComposerId)();
  const dataBindingConfig: IDataBindingConfig = accessStore(sceneComposerId)(dataBindingConfigSelector);
  const valueDataBindingStore = useMemo(() => valueDataBindingProvider.createStore(true), [valueDataBindingProvider]);
  const [dirty, setDirty] = useState(false);
  const [builderState, setBuilderState] = useState<IValueDataBindingProviderState>(
    EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE,
  );

  const filterBuilderState = useCallback(
    (state: IValueDataBindingProviderState) => {
      const bindingTemplateNumOfFields = Object.keys(dataBindingConfig.fieldMapping).length;
      return {
        definitions: state.definitions.slice(0, bindingTemplateNumOfFields),
        selectedOptions: state.selectedOptions.slice(0, bindingTemplateNumOfFields),
      };
    },
    [valueDataBindingProvider],
  );

  useEffect(() => {
    // Subscribe on changes
    valueDataBindingStore.setOnStateChangedListener((state) => {
      setBuilderState(filterBuilderState(state));
    });
    return () => valueDataBindingStore.setOnStateChangedListener(undefined);
  }, [valueDataBindingProvider]);

  const [dataBinding, fieldDisplayNames] = useMemo(() => {
    const fieldDisplayNames: Record<string, string> = {};
    const dataBindingContext = {};
    const dataBinding = { dataBindingContext };

    Object.keys(dataBindingConfig.fieldMapping).forEach((key) => {
      // TwinMaker V1 only uses 1 data binding value
      const templateName = dataBindingConfig.fieldMapping[key][0];
      const templateReference = dataBindingConfig.template?.[templateName];
      fieldDisplayNames[key] = templateName;

      if (templateReference) {
        dataBindingContext[key] = templateReference;
      }
    });

    return [dataBinding, fieldDisplayNames];
  }, [dataBindingConfig]);

  // Save to template variables
  useEffect(() => {
    if (dirty) {
      const newDataBindingConfig: IDataBindingConfig = {
        fieldMapping: cloneDeep(dataBindingConfig.fieldMapping),
        template: {},
      };
      for (let i = 0; i < builderState.definitions.length; ++i) {
        if (!builderState.selectedOptions[i]?.value) {
          break;
        }

        const definition = builderState.definitions[i];
        // TODO: TwinMaker V1 only uses the first mapping
        const templateField = newDataBindingConfig.fieldMapping[definition.fieldName][0];
        const selected = builderState.selectedOptions[i];
        if (selected && selected.value) {
          newDataBindingConfig.template![templateField] = selected.value;
        }
      }

      setSceneProperty(KnownSceneProperty.DataBindingConfig, newDataBindingConfig);
      setDirty(false);
    }
  }, [dirty]);

  useEffect(() => {
    // Initiate the provider
    const state = valueDataBindingStore.setBinding(
      'SceneDataBindingTemplateEditor',
      dataBinding as IValueDataBinding,
      dataBindingConfig,
    );
    setBuilderState(filterBuilderState(state));
  }, [dataBindingConfig, valueDataBindingProvider]);

  return (
    <React.Fragment>
      {builderState.definitions.map((definition, index) => {
        const { options, state } = definition;
        let selectedOption: IDataFieldOption | null = null;
        if (index < builderState.selectedOptions.length) {
          selectedOption = builderState.selectedOptions[index];
        }
        const disabled = state !== 'ready';
        const fieldDisplayName = fieldDisplayNames?.[definition.fieldName] ?? definition.fieldName;

        return (
          <FormField label={fieldDisplayName} key={definition.fieldName}>
            <Select
              data-testid={`data-binding-value-selector-${index}`}
              selectedOption={selectedOption}
              onChange={async (e) => {
                await valueDataBindingStore.updateSelection(definition.fieldName, e.detail.selectedOption);
                setDirty(true);
              }}
              options={options}
              disabled={disabled}
              placeholder={intl.formatMessage({ defaultMessage: 'Select an option', description: 'Section title' })}
            />
          </FormField>
        );
      })}
    </React.Fragment>
  );
};
