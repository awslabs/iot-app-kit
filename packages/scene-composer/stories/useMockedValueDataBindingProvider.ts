import { useCallback, useRef } from 'react';
import { isEqual } from 'lodash';

import {
  IValueDataBindingProvider,
  IDataFieldOption,
  IDataBindingConfig,
  IValueDataBinding,
  IValueDataBindingStore,
  IValueDataBindingProviderState,
} from '../src/interfaces';
import {
  ENTITY_ID_INDEX,
  COMPONENT_NAME_INDEX,
  PROPERTY_NAME_INDEX,
} from '../src/components/panels/scene-components/common/ValueDataBindingBuilder';

import {
  asyncLoadEntityOptions,
  asyncLoadComponentNameOptions,
  asyncLoadPropertyNameOptions,
  cloneStoreState,
  createEmtpyStoreState,
  createIdenticalLabelOption,
  validateEntityId,
} from './useMockedValueDataBindingProviderUtils';
import { FIELD_NAME, MockedValueDataBindingProviderStore } from './types';
import { FIELDS, MOCK_DELAY } from './constants';

export const propertyNames = Object.seal(['temperature']);

export function useMockedValueDataBindingProvider(): IValueDataBindingProvider {
  return {
    useStore: (isDataBindingTemplateProvider: boolean): IValueDataBindingStore => {
      const storeRef = useRef<MockedValueDataBindingProviderStore>({
        state: createEmtpyStoreState(),
        onStateChangedListener: undefined,
      });
      const lastKey = useRef<string>();
      const lastBinding = useRef<IValueDataBinding>();

      const notifyStateChange = useCallback(() => {
        if (storeRef.current.onStateChangedListener) {
          storeRef.current.onStateChangedListener(cloneStoreState(storeRef.current));
        }
      }, []);

      return {
        setBinding(
          key: string,
          binding?: IValueDataBinding,
          dataBindingConfig?: IDataBindingConfig,
        ): IValueDataBindingProviderState {
          // this is an optimization to avoid reinitialize if the state the provider is already synced
          if (lastKey.current !== key || !isEqual(binding, lastBinding.current)) {
            lastKey.current = key;
            lastBinding.current = binding;

            storeRef.current.state = createEmtpyStoreState();

            if (binding) {
              // parse the binding and set to the state storeRef
              const { dataBindingContext } = binding;
              const dataBindingPairs = dataBindingContext as Record<string, string>;
              if (dataBindingPairs) {
                FIELDS.forEach((field) => {
                  if (field in dataBindingPairs) {
                    // TODO: entityId is special as we use the entityName as the label
                    // maybe we should consider put entityName as part of the dataBindingContext
                    // so the select can be initialized with the proper label.
                    // for now, we'll only return the value, and rely on the client to
                    // properly handle the rendering when label is fetched.
                    storeRef.current.state.selectedOptions.push(createIdenticalLabelOption(dataBindingPairs[field]));
                  }
                });
              }
            }

            // initiate loading of entities
            asyncLoadEntityOptions(
              storeRef.current,
              isDataBindingTemplateProvider,
              dataBindingConfig,
              notifyStateChange,
            );
          }

          return cloneStoreState(storeRef.current);
        },

        async updateSelection(
          fieldName: FIELD_NAME,
          selected: IDataFieldOption,
          dataBindingConfig?: IDataBindingConfig,
        ) {
          if (fieldName === 'entityId') {
            try {
              await validateEntityId(selected.value);
              if (storeRef.current.state.errors?.invalidEntityId) {
                storeRef.current.state.errors.invalidEntityId = false;
              }

              storeRef.current.state.selectedOptions[ENTITY_ID_INDEX] = selected;
              // clear rest of the selections
              storeRef.current.state.selectedOptions[COMPONENT_NAME_INDEX] = null;
              storeRef.current.state.selectedOptions[PROPERTY_NAME_INDEX] = null;

              // load component
              asyncLoadComponentNameOptions(
                storeRef.current,
                isDataBindingTemplateProvider,
                dataBindingConfig,
                notifyStateChange,
              );
            } catch (e) {
              if (storeRef.current.state.errors === undefined) {
                storeRef.current.state.errors = {};
              }
              storeRef.current.state.errors.invalidEntityId = true;
              notifyStateChange();
            }
          } else if (fieldName === 'componentName') {
            storeRef.current.state.selectedOptions[COMPONENT_NAME_INDEX] = selected;
            // clear rest of the selections
            storeRef.current.state.selectedOptions[PROPERTY_NAME_INDEX] = null;

            // load property
            asyncLoadPropertyNameOptions(storeRef.current, notifyStateChange);
          } else if (fieldName === 'propertyName') {
            storeRef.current.state.selectedOptions[PROPERTY_NAME_INDEX] = selected;
            notifyStateChange();
          } else {
            // TODO: Match template varialbes
            // const a = 20;
          }
        },

        createBinding(): Promise<IValueDataBinding> {
          return new Promise((resolve) => {
            setTimeout(() => {
              const map = {};
              FIELDS.forEach((field, index) => {
                const selectedOption = storeRef.current.state.selectedOptions[index];
                if (selectedOption) {
                  map[field] = selectedOption?.value;
                }
              });

              const result = {
                dataBindingContext: map,
              };

              lastBinding.current = result;

              resolve(result);
            }, MOCK_DELAY);
          });
        },

        setOnStateChangedListener(listener: (state: IValueDataBindingProviderState) => void | undefined) {
          storeRef.current.onStateChangedListener = listener;
        },
      };
    },
  };
}
