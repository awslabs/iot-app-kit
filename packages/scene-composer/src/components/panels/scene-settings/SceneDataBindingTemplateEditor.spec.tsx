import React from 'react';
import { render } from '@testing-library/react';

import { useMockedValueDataBindingProvider } from '../../../../stories/useMockedValueDataBindingProvider';

import { SceneDataBindingTemplateEditor } from './SceneDataBindingTemplateEditor';

describe('SceneDataBindingTemplateEditor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should change the selection from icon to color', async () => {
    const mockValueDataBindingProvider = useMockedValueDataBindingProvider();
    const { container } = render(
      <SceneDataBindingTemplateEditor valueDataBindingProvider={mockValueDataBindingProvider} />,
    );

    expect(container).toMatchSnapshot();
  });
});
