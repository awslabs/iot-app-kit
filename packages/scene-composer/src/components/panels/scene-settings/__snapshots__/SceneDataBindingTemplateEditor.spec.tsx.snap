// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneDataBindingTemplateEditor should change the selection from icon to color 1`] = `
<div>
  <div
    data-mocked="FormField"
    label="sel_entity"
  >
    <div
      data-mocked="Select"
      data-testid="data-binding-value-selector-0"
      disabled=""
      options="[]"
      placeholder="Select an option"
      selectedoption="null"
    />
  </div>
  <div
    data-mocked="FormField"
    label="sel_comp"
  >
    <div
      data-mocked="Select"
      data-testid="data-binding-value-selector-1"
      disabled=""
      options="[]"
      placeholder="Select an option"
      selectedoption="null"
    />
  </div>
</div>
`;
